liblsp/Server.h

```diff
@@ -0,0 +1,251 @@
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <liblsp/Transport.h>
+#include <liblsp/Range.h>
+
+#include <json/value.h>
+
+#include <functional>
+#include <optional>
+#include <ostream>
+#include <unordered_map>
+#include <variant>
+
+namespace lsp {
+
+class Transport;
+
+enum class ErrorCode;
+
+// {{{ Helper types
+enum class Trace { Off, Messages, Verbose };
+
+struct WorkspaceFolder {
+	std::string name; // The name of the workspace folder. Used to refer to this workspace folder in the user interface.
+	std::string path; // The associated file path for this workspace folder.
+};
+
+struct DocumentPosition {
+	std::string path;
+	Position position;
+};
+
+struct ServerId {
+	std::string serverName;
+	std::string serverVersion;
+};
+
+enum class DocumentHighlightKind {
+	Unspecified,
+	Text,           //!< a textual occurrence
+	Read,           //!< read access to a variable
+	Write,          //!< write access to a variable
+};
+
+struct Location {
+	std::string path;
+	Range range;
+};
+
+struct DocumentHighlight {
+	Range range;
+	DocumentHighlightKind kind = DocumentHighlightKind::Unspecified;
+};
+
+enum class DiagnosticSeverity {
+	Error = 1,
+	Warning = 2,
+	Information = 3,
+	Hint = 4,
+};
+
+enum class DiagnosticTag {
+	Unnecessary = 1, // Unused or unnecessary code.
+	Deprecated = 2   // Deprecated or obsolete code.
+};
+
+/// Represents a related message and source code location for a diagnostic. This should be
+/// used to point to code locations that cause or related to a diagnostics, e.g when duplicating
+/// a symbol in a scope.
+struct DiagnosticRelatedInformation {
+	Location location;   // The location of this related diagnostic information.
+	std::string message; // The message of this related diagnostic information.
+};
+
+/// Represents a diagnostic, such as a compiler error or warning. Diagnostic objects are only valid in the scope of a resource.
+struct Diagnostic {
+	Range range;                                   // The range at which the message applies.
+	std::optional<DiagnosticSeverity> severity;
+	std::optional<unsigned long long> code;        // The diagnostic's code, which might appear in the user interface.
+	std::optional<std::string> source;             // A human-readable string describing the source of this diagnostic, e.g. 'typescript' or 'super lint'.
+	std::string message;                           // The diagnostic's message.
+	std::vector<DiagnosticTag> diagnosticTag;      // Additional metadata about the diagnostic.
+	std::vector<DiagnosticRelatedInformation> relatedInformation; // An array of related diagnostic information, e.g. when symbol-names within a scope collide all definitions can be marked via this property.
+};
+// }}}
+
+/// Solidity Language Server, managing one LSP client.
+///
+/// This implements a subset of LSP version 3.16 that can be found at:
+///     https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/
+class Server
```

Comment:
Do we need this as base class?

Reply:
I'm not 100% insisting. My point however is to split between protocol and actual logic. That is what I yesterday in the meeting meant with "I'd like to keep 2 layers". hmmmmmm... (I'll elaborate).

Reply:
thinking about it, it's basically just an JSON-RPC wrapper, I think I can shift the API in a way we only have JSON-RPC helpers so I can ideally avoid this class.